File: rust/src/day01.rs
use std::collections::HashMap;

fn main() {
    let input = include_str!("../../inputs/day01/input");
    let (mut left_col, mut right_col): (Vec<u32>, Vec<u32>) = input
        .lines()
        .map(|line| {
            let (l, r) = line.split_once("   ").unwrap();
            (l.parse::<u32>().unwrap(), r.parse::<u32>().unwrap())
        })
        .unzip();
    left_col.sort_unstable();
    right_col.sort_unstable();

    println!("part 1: {}", part1(&left_col, &right_col));
    println!("part 2: {}", part2(&left_col, &right_col));
}

fn part1(left_col: &[u32], right_col: &[u32]) -> u32 {
    left_col
        .iter()
        .zip(right_col)
        .map(|(&l, &r)| l.abs_diff(r))
        .sum()
}

fn part2(left_col: &[u32], right_col: &[u32]) -> u32 {
    let counts = right_col.iter().fold(HashMap::new(), |mut counts, &elem| {
        *counts.entry(elem).or_insert(0) += 1;
        counts
    });
    left_col
        .iter()
        .map(|elem| elem * counts.get(elem).unwrap_or(&0))
        .sum()
}

File: zig/day01.zig
const std = @import("std");

pub fn main() !void {
    @setEvalBranchQuota(2_000_000);
    const part1, const part2 = comptime blk: {
        const input = @embedFile("inputs/day01/input");
        const nLines = std.mem.count(u8, input, &[_]u8{'\n'});
        var leftCol: [nLines]u32 = undefined;
        var rightCol: [nLines]u32 = undefined;
        var iter = std.mem.tokenizeAny(u8, input, &std.ascii.whitespace);
        for (0..nLines) |lineIdx| {
            leftCol[lineIdx] = try std.fmt.parseInt(u32, iter.next() orelse @compileError("bad line"), 10);
            rightCol[lineIdx] = try std.fmt.parseInt(u32, iter.next() orelse @compileError("bad line"), 10);
        }
        std.mem.sortUnstable(u32, &leftCol, {}, std.sort.asc(u32));
        std.mem.sortUnstable(u32, &rightCol, {}, std.sort.asc(u32));

        var absDiffSum = 0;
        for (leftCol, rightCol) |l, r| {
            absDiffSum += if (l > r) (l - r) else (r - l);
        }

        var weightedSum = 0;
        for (leftCol) |l| {
            // easy but slow because of quadratic runtime
            weightedSum += l * std.mem.count(u32, &rightCol, &.{l});
        }
        break :blk .{ absDiffSum, weightedSum };
    };
    std.debug.print("part 1: {}\npart2: {}\n", .{ part1, part2 });
}
